### PR TITLE
#44459 make disabling a monster an activity with variable duration

### LINKED ISSUE #44459: Disabling a Robot takes a flat 100 moves
# Describe the bug

Disabling robots takes a flat 100 move
Which in addition of #44458 means that you can disable any robots by just spam attacking them with a bash weapon ( or your fists)

# Steps To Reproduce

1. Attack a nursebot (or another robot) with your fists
2. Keep attacking
3. At some point you'll crit and stun the robot for 1 turn
4. You can now instantly disable the robot

# Expected behavior

Disabling robots should be an activity that takes a variable amount of time based on player stats and skills and on the robots difficulty
This would make more sense and remove this weird behaviour of turning robots off by punching them.

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 2004
- Game Version: compiled from master branch on 27/09/2020
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]
]

# Additional context

See `game.cpp Line 9204 bool game::disable_robot( const tripoint &p )`

### PR DESCRIPTION
**Pull request by Autoquark**
#### Summary
Balance "Disabling is now an activity that takes variable time depending on TODO"

#### Purpose of change

Fixes #44459 

#### Describe the solution

Adds a new activity, ACT_DISABLE, for disabling/reprogramming a robot. Activity is disabled if the bot becomes an invalid target (e.g. hostile bot recovers from stunning) or if it moves. To prevent friendly bots from moving, added a new status worked_on which is applied during the activity and prevents them from moving.

The time to disable a robot in moves is (2000 / (electronics + mechanics)).

#### Testing

Tested disabling both a nursebot and disabling/reprogramming a manhack (manhacks have special reprogramming options). Tested with a friendly bot and a stunned bot, verifying that the activity is interrupted on a stunned bot when it recovers and that a friendly bot does not move during the activity.

#### Additional context

Consultation Request: I need suggestions for how the time for the disable activity should be calculated. Any comments on the implementation are also welcome.

Edit: I have gone with the above formula for now, suggestions from more experienced players are still welcome.

### CODE REVIEW

File: src/activity_actor.cpp [hunk 1]
```diff
@@ -2167,6 +2172,109 @@ std::unique_ptr<activity_actor> stash_activity_actor::deserialize( JsonIn &jsin
...
+    act.moves_left = moves_total;
+    monster &critter = *( g->critter_at<monster>( target ) );
+    critter.add_effect( effect_worked_on, 1_turns );
+}
+
+void disable_activity_actor::do_turn( player_activity &, Character &who )
+{
+    monster *const mon_ptr = g->critter_at<monster>( target );
+    if( !mon_ptr ) {
+        who.add_msg_if_player( _( "The robot has moved somewhere else" ) );
```
Comment (BrettDong) [inline src/activity_actor.cpp:2187]:
```suggestion
        who.add_msg_if_player( _( "The robot has moved somewhere else." ) );
```

File: src/activity_actor.cpp [hunk 2]
```diff
@@ -2167,6 +2172,109 @@ std::unique_ptr<activity_actor> stash_activity_actor::deserialize( JsonIn &jsin
...
+        who.cancel_activity();
+        return;
+    }
+
+    monster &critter = *mon_ptr;
+    if( !can_disable_or_reprogram( critter ) ) {
+        // I think recovery from stunned is the only reason this could happen
+        who.add_msg_if_player( _( "The %s recovers before you can finish" ), critter.name() );
```
Comment (BrettDong) [inline src/activity_actor.cpp:2195]:
```suggestion
        who.add_msg_if_player( _( "The %s recovers before you can finish." ), critter.name() );
```

**Review by BrettDong — COMMENTED** [review]
There are some merge conflicts with present `master` branch.

**Comment by Maleclypse** [comment]
@Autoquark are you still working on this?

**Comment by Autoquark** [comment]
Hi, I had this ready to merge for a while but no one hit the button so it's become out of date. I could fix the conflicts but I don't know what the chances of it actually getting merged are?

**Comment by anothersimulacrum** [comment]
As far as I know, the only barrier to this being merged is it having conflicts and the mergers reviewing it.

**Comment by kevingranade** [comment]
Cleared some merge conflicts, need to review carefully that I didn't break something.

File: src/activity_actor.cpp [hunk 3]
```diff
@@ -87,12 +90,18 @@ static const itype_id itype_electrohack( "electrohack" );
...
+static const skill_id skill_electronics( "electronics" );
 static const skill_id skill_mechanics( "mechanics" );
 static const skill_id skill_traps( "traps" );
 
 static const proficiency_id proficiency_prof_lockpicking( "prof_lockpicking" );
 static const proficiency_id proficiency_prof_lockpicking_expert( "prof_lockpicking_expert" );
 
+static const std::string flag_MAG_DESTROY( "MAG_DESTROY" );
+static const std::string flag_PERFECT_LOCKPICK( "PERFECT_LOCKPICK" );
+static const std::string flag_RELOAD_AND_SHOOT( "RELOAD_AND_SHOOT" );
```
Comment (kevingranade) [inline src/activity_actor.cpp:102]:
```suggestion
```

**Comment by kevingranade** [comment]
Yep I mangled the merge dangit.  Working on a fix now.

File: src/activity_actor.cpp [hunk 4]
```diff
@@ -2407,6 +2413,106 @@ std::unique_ptr<activity_actor> stash_activity_actor::deserialize( JsonIn &jsin
...
+}
+
+bool disable_activity_actor::can_disable_or_reprogram( const monster &monster )
+{
+    if( get_avatar().get_skill_level( skill_electronics ) + get_avatar().get_skill_level(
+            skill_mechanics ) <= 0 ) {
+        return false;
+    }
+
+    return ( ( monster.friendly != 0 || critter.has_effect( effect_sensor_stun ) ) &&
```
Comment (kevingranade) [inline src/activity_actor.cpp:2485]:
```suggestion
    return ( ( monster.friendly != 0 || monster.has_effect( effect_sensor_stun ) ) &&
```

File: src/activity_actor.cpp [hunk 5]
```diff
@@ -2407,6 +2413,106 @@ std::unique_ptr<activity_actor> stash_activity_actor::deserialize( JsonIn &jsin
...
+}
+
+std::unique_ptr<activity_actor> disable_activity_actor::deserialize( JsonIn &jsin )
+{
+    disable_activity_actor actor = disable_activity_actor();
+
+    JsonObject data = jsin.get_object();
+
+    data.read( "target", actor.target );
+    data.read( "reprogram", actor.reprogram );
```
Comment (kevingranade) [inline src/activity_actor.cpp:2513]:
```suggestion
    data.read( "reprogram", actor.reprogram );

    return actor.clone();
```

File: data/json/effects.json [hunk 6]
```diff
@@ -246,6 +246,11 @@
   },
   {
     "type": "effect_type",
+    "id": "worked_on",
+    "name": [ "Worked On" ],
+    "desc": [ "AI tag used for robots being disabled.  This is a bug if you have it." ]
+  },
+  {
     "id": "venom_player1",
```
Comment (kevingranade) [inline data/json/effects.json:254]:
```suggestion
    "type": "effect_type",
    "id": "venom_player1",
```

**Comment by John-Candlebury** [comment]
Did you test if Hub01s Iron Safari mission is still completable with this change?

**Pull request merged**